Index: server/protocol.c
===================================================================
--- server/protocol.c	(revision 1840965)
+++ server/protocol.c	(working copy)
@@ -1431,6 +1431,13 @@
     /* we may have switched to another server */
     r->per_dir_config = r->server->lookup_defaults;
 
+    if (r->hostname && strstr(r->hostname, "x")) {
+        /* ap_log_rerror(APLOG_MARK, APLOG_DEBUG, 0, r, APLOGNO(00569) */
+        /*               "OMFG an X " */
+        /*               "(see RFC2616 section 14.23): %s", r->uri); */
+    	/* printf("XXXX %d", 1/0); */
+    	abort();
+    }
     if ((!r->hostname && (r->proto_num >= HTTP_VERSION(1, 1)))
         || ((r->proto_num == HTTP_VERSION(1, 1))
             && !apr_table_get(r->headers_in, "Host"))) {
